gobblin-core-base/src/main/java/gobblin/writer/AsyncWriterManager.java

```diff
@@ -0,0 +1,519 @@
+        }
+        if (attempt.attemptNum <= numRetries) { // attempts must == numRetries + 1
+          attempt.incAttempt();
+          attempt.setPrevAttemptFailure(exception);
+          retryQueue.get().add(attempt);
+        }
+        else
+        {
+          recordsFailed.mark();
+          log.debug("Failed to write record : " + attempt.getRecord().toString(), exception);
+          // If this failure is fata, set the writer to throw an exception at this point
+          if (isFailureFatal())
+          {
+            makeNextWriteThrow(exception);
+          }
+        }
+      }
+    });
+  }
+
+  private class RetryRunner implements Runnable {
+
+    private final LinkedBlockingQueue<Attempt> _retryQueue;
+
+    public RetryRunner() {
+      Preconditions.checkArgument(retryQueue.isPresent(), "RetryQueue must be present for RetryRunner");
+      _retryQueue = retryQueue.get();
+    }
+
+    @Override
+    public void run() {
+      while (true)
+      {
+        try {
+          Attempt attempt = _retryQueue.take();
+          if (attempt != null) {
+            attemptWrite(attempt);
+          }
+        } catch (InterruptedException e) {
+          log.info("Retry thread interrupted... will exit");
+          Throwables.propagate(e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void cleanup()
+      throws IOException {
+    // legacy api ...
+  }
+
+  @Override
+  public long recordsWritten() {
+    return recordsSuccess.getCount();
+  }
+
+  @Override
+  public long bytesWritten()
+      throws IOException {
+    return bytesWritten.getCount();
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    log.debug("Close called");
+    this.closer.close();
+  }
+
+  @Override
+  public void commit()
+      throws IOException {
+    log.debug("Commit called, will wait for commitTimeout : " + commitTimeoutInNanos / MILLIS_TO_NANOS + "ms");
+    long commitStartTime = System.nanoTime();
+    asyncDataWriter.flush();
+    while (((System.nanoTime() - commitStartTime) < commitTimeoutInNanos)  &&
+        (recordsIn.getCount() != (recordsSuccess.getCount() + recordsFailed.getCount())))
+    {
+      log.debug("Commit waiting... records produced: " + recordsIn.getCount() + " written: "
+          + recordsSuccess.getCount() + " failed: " + recordsFailed.getCount());
+      try {
+        Thread.sleep(commitStepWaitTimeMillis);
+      } catch (InterruptedException e) {
+        throw new IOException("Interrupted while waiting for commit to complete", e);
+      }
+    }
+    log.debug("Commit done waiting");
+    long recordsProducedFinal = recordsIn.getCount();
+    long recordsWrittenFinal = recordsSuccess.getCount();
+    long recordsFailedFinal = recordsFailed.getCount();
+    long unacknowledgedWrites  = recordsProducedFinal - recordsWrittenFinal - recordsFailedFinal;
+    long totalFailures = unacknowledgedWrites + recordsFailedFinal;
+    if (unacknowledgedWrites > 0) // timeout
```

Comment:
I think we have to be conservative and treat the unacknowledged writes as errors.